### PR TITLE
[Ready to Merge] First pass updating with reaction documentation.

### PR DESCRIPTION
**Pull request by pjoviak**
[https://github.ibm.com/toscana/planning/issues/33817](https://github.ibm.com/toscana/planning/issues/33817) Add documentation for reactions

Please feel free to suggest any changes, especially in the wording. Thanks!

@roy713@miguelestrada @jarusso

### CODE REVIEW

**Comment by pjoviak** [comment]
@roy713 Could you please review?

File: guides/V1_Add_Reaction.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+## Example Request
+
+~~~~
+Method: POST
+URL: https://api.watsonwork.ibm.com/graphql
+Headers: 'Content-Type: application/graphql' , 'x-graphql-view: PUBLIC, EXPERIMENTAL'
+Body:
+{
+  mutation {
+    addReaction(input: {targetId: "5a209f74e4b0faa757ac1afd", reaction: "?"}) {
```
Comment (roy713) [inline guides/V1_Add_Reaction.md]: Maybe we can add an emoji here as the reaction to make it a bit more descriptive?

File: guides/V1_Reacting_Users.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+## Example Request
+
+~~~~
+Method: POST
+URL: https://api.watsonwork.ibm.com/graphql
+Headers: 'Content-Type: application/graphql' , 'x-graphql-view: PUBLIC, EXPERIMENTAL'
+Body:
+{
+  query {
+    reactingUsers(targetId: "5a9ee86ee4b0cb1ca3dfef33", reaction: "?") {
```
Comment (roy713) [inline guides/V1_Reacting_Users.md]: Same comment as above

File: guides/V1_Remove_Reaction.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+## Example Request
+
+~~~~
+Method: POST
+URL: https://api.watsonwork.ibm.com/graphql
+Headers: 'Content-Type: application/graphql' , 'x-graphql-view: PUBLIC, EXPERIMENTAL'
+Body:
+{
+  mutation {
+    removeReaction(input: {targetId: "5a209f74e4b0faa757ac1afd", reaction: "?"}) {
```
Comment (roy713) [inline guides/V1_Remove_Reaction.md]: Same comment as above

File: guides/V1_Remove_Reaction.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
+---
+copyright: 'Copyright IBM Corp. 2018'
+link: 'remove-a-reaction'
+is: 'future'
```
Comment (roy713) [inline guides/V1_Remove_Reaction.md]: Unclear if, since we're under the `EXPERIMENTAL` header, we should mark this as `experimental` instead of `future`. Same goes for all the others but probably @miguelestrada has a better idea.

File: guides/V1_Add_Reaction.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+copyright: 'Copyright IBM Corp. 2018'
+link: 'add-a-reaction'
+is: 'future'
+---
+
+# Add a Reaction
+
+## Concepts
+
+The _addReaction_ mutation allows the user to add a reaction to a message.  The mutation accepts a Message ID (targetId) and a reaction as arguments, and makes the request on behalf of the calling user, adding the specified reaction to the message.
```
Comment (roy713) [inline guides/V1_Add_Reaction.md]: Maybe we can add to the description for both mutations how the result contains an updated count and whether the user has reacted or not (probably worded better than that 😆).

File: guides/V1_message_main.md [hunk 6]
```diff
@@ -22,3 +22,4 @@ Watson Work Services represents a Message with this object.
 | content   | String | The body of the message |
 | contentType | String | Mime type of the message content  |
 | annotations | [String] | A set of optional objects/attachments that are added to a message, called annotations. Each annotation represents meta data that are related to the message. An annotation is represented by a structured format, which can be viewed as a template. There are different annotation types, some are used to store results of cognitive analysis to a message for example. An annotation can be added at message create time or later, as an update.  Apps are currently limited to create only one type of annotation called `generic` |
+| reactions | [Reaction] | A set of optional objects that are added to a message, called reactions. Each reaction represents how users reacted (by emoji) to the message. A Reaction is represented by the reaction (String emoji), the count (Int number of users that reacted with the same reaction), and viewerHasReacted (Boolean true the calling user reacted with given reaction, or false didn not). This requires the `EXPERIMENTAL` value in `x-graphql-view header`. |
```
Comment (roy713) [inline guides/V1_message_main.md]: You can ignore my comment regarding the description for the mutations (adding count and user has reacted) since its explained here although I don't think it hurts having it in both places.

Also there seems to be a typo:
> (Boolean true the calling user reacted with given reaction, or false didn not).

File: guides/V1_wwsg_Webhooks.md [hunk 7]
```diff
@@ -43,6 +43,10 @@ Available events are:
 
 **message-annotation-removed**: each time an annotation is removed from a message
 
+**reaction_added**: each time a reaction is added to a message
+
+**reaction_removed**: each time a reaction is removed to a message
```
Comment (roy713) [inline guides/V1_wwsg_Webhooks.md]: `removed from a message`?

File: references/V1_OutboundCallback.yml [hunk 8]
```diff
@@ -334,6 +338,84 @@ definitions:
...
+      time:
+        description: "Time and date of the reaction being added, in milliseconds since January 1st, 00:00, 1970 UTC"
+        type: integer
+        format: int64
+      reaction:
+        description: "The reaction that was added."
+        type: string
+      userId:
+        description: "Id of the user that added the reaction."
+        type: string
```
Comment (roy713) [inline references/V1_OutboundCallback.yml]: Some descriptions have `"` and some don't - we may want to keep consistent (haven't looked at what's the common behavior in others).

File: references/V1_OutboundCallback.yml [hunk 9]
```diff
@@ -334,6 +338,84 @@ definitions:
...
+      - and belong to an app that is a member of the space.
+    properties:
+      type:
+        description: The event type is `reaction-removed`.
+        type: string
+      spaceId:
+        description: Id of the space in which the reaction was removed.
+        type: string
+      objectId:
+        description: Id of the object being reacted to. Currently, only messages can be reacted to..
```
Comment (roy713) [inline references/V1_OutboundCallback.yml]: `..`?

File: references/V1_OutboundCallback.yml [hunk 10]
```diff
@@ -334,6 +338,84 @@ definitions:
...
+        type: string
+      time:
+        description: "Time and date of the reaction being removed, in milliseconds since January 1st, 00:00, 1970 UTC"
+        type: integer
+        format: int64
+      reaction:
+        description: "The reaction that was removed."
+        type: string
+      userId:
+        description: "Id of the user that removed the reaction."
```
Comment (roy713) [inline references/V1_OutboundCallback.yml]: Same comment as above for the `"`

File: references/V1_OutboundCallback.yml [hunk 11]
```diff
@@ -484,3 +566,4 @@ paths:
             Retries are made at time intervals of every 30s during the first 2 hours and then at 3h, 6h, 12h, 24h, 36h, 72h after the first failed notification.
             After this, the event will be discarded.
             This retry is tied to the space for the event in question.
+
```
Comment (roy713) [inline references/V1_OutboundCallback.yml]: Line ending? 🤣

File: guides/V1_Remove_Reaction.md (hunk 3 above)
Comment (pjoviak) [reply guides/V1_Remove_Reaction.md]: Can't do to file code page cp1252

File: guides/V1_Add_Reaction.md (hunk 1 above)
Comment (pjoviak) [reply guides/V1_Add_Reaction.md]: Can't do to file code page cp1252

File: guides/V1_Reacting_Users.md (hunk 2 above)
Comment (pjoviak) [reply guides/V1_Reacting_Users.md]: Can't do to file code page cp1252

**Comment by pjoviak** [comment]
@roy713 @jarusso @miguelestrada   Suggested changes have been made.  Please review, and merge if there are no further comments or suggestions.  Thank you.

**Review by roy713 — APPROVED** [review]
LGTM 👍

**Review by gary-mguo — APPROVED** [review]
LGTM 👍

**Comment by pjoviak** [comment]
@jarusso @miguelestrada  Please review, and merge if there are no further comments or suggestions. Thank you.

**Comment by jarusso** [comment]
these look ok to me - one thought though, we have a set of detailed items here around reactions - we need an overall introduction to the concept of reactions (at a more meta level) and then have these detailed items about reactions.

File: guides/V1_Reacting_Users.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+copyright: 'Copyright IBM Corp. 2018'
+link: 'reacting-users'
+is: 'experimental'
+---
+
+# Query for Users Reacting to a Message
+
+## Concepts
+
+The _reactingUsers_ query allows the user to get all users who reacted to a message with a specified reaction.  The query accepts a Message ID (targetId) and a reaction as arguments, and optional paging information.
```
Comment (miguelestrada) [inline guides/V1_Reacting_Users.md]: Can we change 

>allows the user

to:

>allows the API caller

as it may be an app/bot or service that is making the API call....

File: guides/V1_Reacting_Users.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+type ReactingUserCollection {
+  totalCount: Int
+  pageInfo: PageInfo!
+  items: [ReactingUser]!
+}
+
+```
+
+## Example Request
```
Comment (miguelestrada) [inline guides/V1_Reacting_Users.md:38]: In general, when showing example requests it is useful to also show the response... it helps comparing to what developers may be seeing on their end.

Won't hold the PR for this, but may want to consider subsequent PR to add sample output

File: guides/V1_Add_Reaction.md [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+copyright: 'Copyright IBM Corp. 2018'
+link: 'add-a-reaction'
+is: 'experimental'
+---
+
+# Add a Reaction
+
+## Concepts
+
+The _addReaction_ mutation allows the user to add a reaction to a message.  The mutation accepts a Message ID (targetId) and a reaction as arguments, and makes the request on behalf of the calling user, adding the specified reaction to the message.  The result will reflect the current **count** of users that reacted with the same **reaction** (emoji), and **viewerHasReacted** as to whether the calling user has reacted (true).
```
Comment (miguelestrada) [inline guides/V1_Add_Reaction.md]: Can we change 

>allows the user

to:

>allows the API caller

as it may be an app/bot or service that is making the API call....

File: guides/V1_Remove_Reaction.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+copyright: 'Copyright IBM Corp. 2018'
+link: 'remove-a-reaction'
+is: 'experimental'
+---
+
+# Remove a Reaction
+
+## Concepts
+
+The _removeReaction_ mutation allows the user to remove a reaction from a message.  The mutation accepts a Message ID (targetId) and a reaction as arguments, and makes the request on behalf of the calling user, removing the specified reaction from the message.  The result will reflect the current **count** of users that reacted with the same **reaction** (emoji), and **viewerHasReacted** as to whether the calling user has reacted (false).
```
Comment (miguelestrada) [inline guides/V1_Remove_Reaction.md]: Can we change 

>allows the user

to:

>allows the API caller

as it may be an app/bot or service that is making the API call....

File: guides/V1_Reacting_Users.md (hunk 13 above)
Comment (miguelestrada) [reply guides/V1_Reacting_Users.md:38]: Can the reaction utilized in all the samples be `":-)"` instead of `"?"` ?
Alternatively, we can use `"\uD83D\uDC4D"` which is the thumbs up although I rather use that in a separate doc (yet to be written) explaining the relationship between emoji's and reactions @jarusso FYI

File: guides/V1_get_message.md [hunk 16]
```diff
@@ -14,6 +14,7 @@ query getMessage {
     content
     contentType
     annotations
+    reactions
```
Comment (miguelestrada) [inline guides/V1_get_message.md]: We don't need reactions added here.  This document is an intro on how to get info about a message and does not necessarily has to get all the info....   Suggest to drop this file change

File: guides/V1_message_main.md [hunk 17]
```diff
@@ -22,3 +22,4 @@ Watson Work Services represents a Message with this object.
 | content   | String | The body of the message |
 | contentType | String | Mime type of the message content  |
 | annotations | [String] | A set of optional objects/attachments that are added to a message, called annotations. Each annotation represents meta data that are related to the message. An annotation is represented by a structured format, which can be viewed as a template. There are different annotation types, some are used to store results of cognitive analysis to a message for example. An annotation can be added at message create time or later, as an update.  Apps are currently limited to create only one type of annotation called `generic` |
+| reactions | [Reaction] | A set of optional objects that are added to a message, called reactions. Each reaction represents how users reacted (by emoji) to the message. A Reaction is represented by the reaction (String emoji), the count (Int number of users that reacted with the same reaction), and viewerHasReacted (Boolean: true, the calling user reacted with given reaction, or false, the calling user did not). This requires the `EXPERIMENTAL` value in `x-graphql-view header`. |
```
Comment (miguelestrada) [inline guides/V1_message_main.md]: - remove "(by emoji)"
- change "(String emoji)" to "(String, usually an emoji but not required)"

Rationale:  Reaction APIs allow for the association of "reactions" to objects such as "messages".   A "reaction" is represented simply by a String.  The Workspace app happens to utilize a particular set Emojis to represent the reactions.  This separation of Emoji from Reactions allow for future implementations and for other clients to choose different set of Emojis or their own custom set of Strings.  The main requirement is that it be a String.

File: guides/V1_wwsg_Webhooks.md [hunk 18]
```diff
@@ -43,6 +43,10 @@ Available events are:
 
 **message-annotation-removed**: each time an annotation is removed from a message
 
+**reaction_added**: each time a reaction is added to a message
```
Comment (miguelestrada) [inline guides/V1_wwsg_Webhooks.md]: Add "(API is EXPERIMENTAL)" at the end of each of these

File: references/V1_OutboundCallback.md [hunk 19]
```diff
@@ -265,6 +267,24 @@ Send back the verification challenge to show that the webhook supports this call
 |**response**  <br>*required*|The challenge that was provided in the request body of the verification request.|string|
 
 
+<a name="addreactionbody"></a>
+### AddReactionBody
```
Comment (miguelestrada) [inline references/V1_OutboundCallback.md:271]: Can these be formatted in the same way (using a table) as the rest of the other data structures for events?
Seems that the actual elements of the structure are missing.  @nivekselig I believe you will be replacing this file based on output of build which is based on the .yml counterpart.  IOW, this .md file is build output and should not be part of this PR.

**Review by miguelestrada — CHANGES_REQUESTED** [review]

File: references/V1_OutboundCallback.md (hunk 19 above)
Comment (pjoviak) [reply references/V1_OutboundCallback.md:271]: Left this since it will get regenerated when build is done, anyway.

**Comment by pjoviak** [comment]
@miguelestrada  @jarusso 
I believe I have addressed all concerns.  Please review and merge.
Thank you.
FYI: @garymg  @roy713

**Comment by pjoviak** [comment]
@miguelestrada @jarusso
I believe I have addressed all concerns. Please review and merge.
Thank you.

File: NavigationMenu.md [hunk 20]
```diff
@@ -33,6 +33,12 @@
   * [Update space membership](./guides/V1_update_space.md)
 
 
+* ## [Reactions](./guides/V1_wwsg_Reactions.md)
```
Comment (miguelestrada) [inline NavigationMenu.md]: NavigationMenu.md should not be changed unless API is `BETA`

Comment (jarusso) [inline NavigationMenu.md]: since this is experimental, it should not go in the NavigationMenu file, can you remove this for now @pjoviak

File: guides/V1_wwsg_Reactions.md [hunk 21]
```diff
@@ -0,0 +1,35 @@
+---
+copyright: 'Copyright IBM Corp. 2018'
+link: 'reactions-overview'
+is: 'experimental'
+---
+## Reactions Overview
+
+Watson Work Services provides a set of **reaction** APIs for callers to interact with **messages**, adding and removing reactions associated with a message.
+
+Reaction APIs allow for the association of "reactions" to objects such as "messages". A "reaction" is represented simply by a String. The Workspace application utilizes a particular set of emojis to represent the reactions. This separation of emoji from reactions allow for future implementations and for other clients to choose different set of emojis or their own custom set of Strings. The main requirement is the reaction is a String.
```
Comment (jarusso) [inline guides/V1_wwsg_Reactions.md]: _this_

> The main requirement is the reaction is a String.

_should be removed, it's unnecessary_

**Review by jarusso — CHANGES_REQUESTED** [review]

**Comment by jarusso** [comment]
@pjoviak ☝️

**Comment by pjoviak** [comment]
@jarusso  Done.

**Review by jarusso — APPROVED** [review]

**Comment by pjoviak** [comment]
@miguelestrada please review, approve and merge.  Thank you.

File: guides/V1_wwsg_Reactions.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+Reaction APIs allow for the association of "reactions" to objects such as "messages". A "reaction" is represented simply by a String. The Workspace application utilizes a particular set of emojis to represent the reactions. This separation of emoji from reactions allow for future implementations and for other clients to choose different set of emojis or their own custom set of Strings.
+
+![Image](../images/reaction.png)
+
+Reactions are a quick way to respond to someone's message.  The caller can select any reaction (string/emoji) of their preference, and associate it with a message.  The message object will optionally retrieve all the reactions to a message, returning each **reaction** (string/emoji), the current number (**count**) of reactors for each of those reactions and if the caller has already reacted or not (**viewerHasReacted**).  Separately, the caller can get a paginated list of users who reacted with a specific reaction.
+
+### External APIs
+
+ * GraphQL mutation for adding reaction
+    * addReaction [Add a reaction to a message](./guides/V1_Add_Reaction.md)
```
Comment (miguelestrada) [inline guides/V1_wwsg_Reactions.md]: from:
> addReaction

to:

>`**addReaction**`

to match the others?

**Review by miguelestrada — CHANGES_REQUESTED** [review]
Minor typo found to bold addReaction

Also, was the webhook yml file updated to specify `EXPERIMENTAL` classification for the two events?

File: references/V1_OutboundCallback.yml [hunk 23]
```diff
@@ -334,6 +338,84 @@ definitions:
       - messageId
       - annotationId
       - time
+  AddReactionBody:
+    type: object
+    description: |
+      Notifies the addition of an emoji reaction to a message.
```
Comment (miguelestrada) [inline references/V1_OutboundCallback.yml]: From:

> Notifies the addition of an emoji reaction to a message.

To:

> Notifies the addition of a reaction to a message.  Note that this event is currently classified as an EXPERIMENTAL API.

File: references/V1_OutboundCallback.yml [hunk 24]
```diff
@@ -334,6 +338,84 @@ definitions:
...
+      - spaceId
+      - objectId
+      - objectType
+      - time
+      - reaction
+      - userId
+  RemoveReactionBody:
+    type: object
+    description: |
+      Notifies the removal of an emoji reaction from a message.
```
Comment (miguelestrada) [inline references/V1_OutboundCallback.yml]: From:

> Notifies the removal of an emoji reaction from a message.

To:

> Notifies the addition of a reaction to a message.  Note that this event is currently classified as an EXPERIMENTAL API.

**Review by miguelestrada — APPROVED** [review]
